### PR TITLE
add share by mail activity

### PR DESCRIPTION
**Pull request by schiessle**
Add share by mail activity. If a file shared by mail gets downloaded the user will see this activity:

activity

Following PR introduces the "email" formatter to the activity app: https://github.com/nextcloud/activity/pull/63

cc @nickvergessen  @rullzer

### CODE REVIEW

**Comment by nickvergessen** [comment]
Can you also add it to https://github.com/nextcloud/server/blob/master/lib/public/RichObjectStrings/definitions.json
Because that is what we are moving to in the long term with activities as well.

**Comment by schiessle** [comment]
@nickvergessen see https://github.com/nextcloud/server/pull/1998/commits/8ef5b0aac3ebfa430e997f6a6d2581836955ca2e

**Comment by MorrisJobke** [comment]
@schiessle How can I share by email? The normal share input only gives me the option to create a remote share if I enter an email address.

**Comment by schiessle** [comment]
if the "share by mail" app is enabled and you enter a valid email address (something like "user@localhost") doesn't work it should show up in the auto-complete list

**Comment by nickvergessen** [comment]
> How can I share by email? The normal share input only gives me the option to create a remote share if I enter an email address.

Did you enable the app?

**Comment by MorrisJobke** [comment]
> Did you enable the app?

No 😕

File: lib/public/RichObjectStrings/definitions.json [hunk 1]
```diff
@@ -224,5 +224,43 @@
         "example": "Support Team"
       }
     }
+  },
+  "file_shared_with_email_downloaded": {
```
Comment (nickvergessen) [inline lib/public/RichObjectStrings/definitions.json]: The object you want to describe here is "email".
E.g.: https://gist.github.com/nickvergessen/eba3e17f2f5cd28831423eadd52763c3

**Review by nickvergessen — DISMISSED** [review]

File: lib/public/RichObjectStrings/definitions.json (hunk 1 above)
Comment (schiessle) [inline lib/public/RichObjectStrings/definitions.json]: updated

**Comment by MorrisJobke** [comment]
Tested and works 👍

**Comment by nickvergessen** [comment]
Btw shouldn't this be part of the sharebymail app?

File: apps/files_sharing/lib/Activity.php [hunk 2]
```diff
@@ -347,10 +364,11 @@ public function getSpecialParameterList($app, $text) {
 						1 => 'username',
 						2 => '',
 					];
-				case self::SUBJECT_SHARED_EMAIL:
```
Comment (nickvergessen) [inline apps/files_sharing/lib/Activity.php]: This is not intentional?

Comment (schiessle) [inline apps/files_sharing/lib/Activity.php]: moved to the share-by-mail app, see my last cmmit

File: apps/files_sharing/lib/Activity.php [hunk 3]
```diff
@@ -114,6 +117,7 @@ public function getNotificationTypes($languageCode) {
 			self::TYPE_SHARED => (string) $l->t('A file or folder has been <strong>shared</strong>'),
 			self::TYPE_REMOTE_SHARE => (string) $l->t('A file or folder was shared from <strong>another server</strong>'),
 			self::TYPE_PUBLIC_LINKS => (string) $l->t('A public shared file or folder was <strong>downloaded</strong>'),
+			self::TYPE_EMAIL_SHARE => (string) $l->t('A file or folder shared by mail was <strong>downloaded</strong>'),
```
Comment (nickvergessen) [inline apps/files_sharing/lib/Activity.php]: Could use the same setting as above (less options = better?)

**Comment by schiessle** [comment]
@nickvergessen I cleaned up the activity types... please have a final look. Thanks!

File: apps/sharebymail/lib/Activity.php [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+	 * @param array $activity
+	 * @return integer|false
+	 * @since 8.0.0
+	 */
+	public function getGroupParameter($activity) {
+		if ($activity['app'] === self::SHARE_BY_MAIL_APP) {
+			switch ($activity['subject']) {
+				case self::SUBJECT_SHARED_EMAIL_BY:
+					// Group by file name
+					return 0;
```
Comment (nickvergessen) [inline apps/sharebymail/lib/Activity.php]: Should return 1 here as well, will fix

**Comment by nickvergessen** [comment]
Fixed two typos, but works now 👍

**Comment by schiessle** [comment]
failing tests are unrelated, all tests passed before the last commit which just fixes two typos.

**Pull request merged**